src/libraries/System.Formats.Tar/tests/TarEntry/TarEntryPax.Tests.cs

```diff
@@ -36,6 +36,122 @@ public void Constructor_UnsupportedEntryTypes()
+        {
+            using MemoryStream source = GetTarMemoryStream(CompressionMethod.Uncompressed, TestTarFormat.v7, "file");
+            using WrappedStream wrappedSource = new WrappedStream(source, canRead: true, canWrite: false, canSeek: false);
+
+            using TarReader sourceReader = new TarReader(wrappedSource, leaveOpen: true);
+            V7TarEntry v7Entry = sourceReader.GetNextEntry(copyData: false) as V7TarEntry;
+            PaxTarEntry paxEntry = new PaxTarEntry(other: v7Entry); // Convert, and avoid advancing wrappedSource position
+
+            using MemoryStream destination = new MemoryStream();
+            using (TarWriter writer = new TarWriter(destination, TarEntryFormat.Pax, leaveOpen: true))
+            {
+                writer.WriteEntry(paxEntry); // Write DataStream exactly where the wrappedSource position was left
+            }
+
+            destination.Position = 0; // Rewind
+            using (TarReader destinationReader = new TarReader(destination, leaveOpen: false))
+            {
+                PaxTarEntry resultEntry = destinationReader.GetNextEntry() as PaxTarEntry;
+                Assert.NotNull(resultEntry);
+                using (StreamReader streamReader = new StreamReader(resultEntry.DataStream))
+                {
+                    Assert.Equal("Hello file", streamReader.ReadToEnd());
+                }
+            }
+        }
+
+        [Fact]
+        public void Constructor_ConversionFromUstar_From_UnseekableTarReader()
+        {
+            using MemoryStream source = GetTarMemoryStream(CompressionMethod.Uncompressed, TestTarFormat.ustar, "file");
+            using WrappedStream wrappedSource = new WrappedStream(source, canRead: true, canWrite: false, canSeek: false);
+
+            using TarReader sourceReader = new TarReader(wrappedSource, leaveOpen: true);
+            UstarTarEntry ustarEntry = sourceReader.GetNextEntry(copyData: false) as UstarTarEntry;
+            PaxTarEntry paxEntry = new PaxTarEntry(other: ustarEntry); // Convert, and avoid advancing wrappedSource position
+
+            using MemoryStream destination = new MemoryStream();
+            using (TarWriter writer = new TarWriter(destination, TarEntryFormat.Pax, leaveOpen: true))
+            {
+                writer.WriteEntry(paxEntry); // Write DataStream exactly where the wrappedSource position was left
+            }
+
+            destination.Position = 0; // Rewind
+            using (TarReader destinationReader = new TarReader(destination, leaveOpen: false))
+            {
+                PaxTarEntry resultEntry = destinationReader.GetNextEntry() as PaxTarEntry;
+                Assert.NotNull(resultEntry);
+                using (StreamReader streamReader = new StreamReader(resultEntry.DataStream))
+                {
+                    Assert.Equal("Hello file", streamReader.ReadToEnd());
+                }
+            }
+        }
```

Comment:
All of these look like they could be driven by one generic driver, rather than be copy pasted a lot.

```C#
        public void Constructor_ConversionFromUstar_From_UnseekableTarReader()
        {
            Constructor_ConversionFrom_UnseekableTarReader(
                TestTarFormat.ustar,
                static entry => new PaxTarEntry(entry));
        }

        private static Type EntryTypeFromFormat(TestTarFormat format)
        {
           return format switch
           {
               TestTarFormat.ustar => typeof(UstarTarEntry),
               ...
           };
        }

        private static Type FormatFromEntryType(Type entryType)
        {
            return entryType switch
            {
                typeof(PaxTarEntry) => TarEntryFormat.Pax,
                ...
            };
        }

        private void Constructor_ConversionFrom_UnseekableTarReader<TTo>(TestTarFormat format, Func<TarEntry,TTo> converter)
        {
            using MemoryStream source = GetTarMemoryStream(CompressionMethod.Uncompressed, TestTarFormat.ustar, "file");
            using WrappedStream wrappedSource = new WrappedStream(source, canRead: true, canWrite: false, canSeek: false);

            using TarReader sourceReader = new TarReader(wrappedSource, leaveOpen: true);
            TarEntry entry = sourceReader.GetNextEntry(copyData: false);
            Assert.IsType(EntryTypeFromFormat(format), entry);
            TTo newEntry = converter(entry); // Convert, and avoid advancing wrappedSource position

            using MemoryStream destination = new MemoryStream();
            using (TarWriter writer = new TarWriter(destination, FormatFromEntryType(typeof(TTo)), leaveOpen: true))
            {
                writer.WriteEntry(newEntry); // Write DataStream exactly where the wrappedSource position was left
            }

            destination.Position = 0; // Rewind
            using (TarReader destinationReader = new TarReader(destination, leaveOpen: false))
            {
                TarEntry resultEntry = destinationReader.GetNextEntry();
                Assert.NotNull(resultEntry);
                Assert.IsType<TTo>(resultEntry);
                using (StreamReader streamReader = new StreamReader(resultEntry.DataStream))
                {
                    Assert.Equal("Hello file", streamReader.ReadToEnd());
                }
            }
        }
```

Reply:
While I appreciate the reduction of code, I think it makes it harder to understand at first glance.

Reply:
Looking at this again, I can't help but make the same suggestion.  There's so much code in these methods that it's hard to tell what's going on.  Reading one nicely written driver method that then just gets invoked multiple times would make it easier (to my way of thinking) to see that the full matrix is being completed; and not wasting time looking to see if the (e.g.) GNU -> Ustar test is subtly different than the PAX -> Ustar.  (Not to mention the number of places that you'd have to update if you want to add an assertion or otherwise enhance/change the test).

Even just taking the middle segment out and making something like

```C#
private static MemoryStream ConvertTarFormat(Stream source, TarEntryFormat targetFormat, TarEntryFormat? sourceFormat = null)
{
    MemoryStream destination = new MemoryStream();

    using (TarReader reader = new TarReader(source, leaveOpen: true /* if you want */))
    using (TarWriter writer = new TarWriter(destination, leaveOpen: true))
    {
        TarEntry? entry = reader.GetNextEntry(copyData: false);

        while (entry is not null)
        {
            if (sourceFormat.HasData)
            {
                Assert.Equals(sourceFormat.Value, entry.Format);
            }

            writer.WriteEntry(entry);
            entry = reader.GetNextEntry(copyData: false);
        }
    }

    // Rewind the stream so it's ready to read.
    destination.Position = 0;
    return destination;
}
```

would help take a lot of the copy-pasta and visual drain out of these tests.

Reply:
I stand corrected. I recycled the code and it looks better. 🙂